Wrap error from embedding files to prevent erasure of the manifest

Nirusu opened the pull request:
### Proposed changes
- Wrap error from embedding files to prevent erasure of the manifest

This fixes a bug caused when trying to embed a file that does not exist on the host filesystem when calling `ego sign <binary-name>` specifically. It does not occur when calling just `ego sign` or `ego sign enclave.json`.

### Related issue
Discovered this while working on #178

Let's take a look at the implementation of the command for context:
https://github.com/edgelesssys/ego/blob/908ace6f269b62fd0b5134f81da0738badac7267/ego/cli/sign.go#L178-L195

In case no filename or a .json file is specified, the EGo CLI first calls `readConfigJSONtoStruct` and then `signWithJSON`, while when a specific (binary) filename is specified, the EGo CLI calls `signExecutable`.

SignExecutable also calls `readConfigJSONToStruct`, but specifically ignores the "no such files or directory" error by checking `!os.IsNotExist(err)`:
https://github.com/edgelesssys/ego/blob/908ace6f269b62fd0b5134f81da0738badac7267/ego/cli/sign.go#L97-L108

The intention here is to create a new `enclave.json` in case none does exist. 

However, since embedding of the files also happens in this function which also spits out the same "no such file or directory" error in case one is missing, we might accidentally overwrite the manifest with the default template as we handle it with the same error check and _think_ the manifest does not exist, while it's just the files to be embedded that are not at the expected path.

## Proposed solution
While I personally find the call structure a bit confusing and would think of refactoring at some point in the future, the easy fix would be to just wrap the error from embedding the files. That's what this PR does, plus adding some test cases for the unit test to handle all the different call paths for the embedding files test.

thomasten commented on ego/cli/sign.go:
```diff
@@ -152,12 +152,13 @@ func (c *Cli) readConfigJSONtoStruct(path string) (*config.Config, error) {
 	if err := conf.PopulateContent(c.fs); err != nil {
-		return nil, err
+		// Wrap error since certain callers may ignore os.IsNotExist to create a new manifest
+		return nil, fmt.Errorf("failed to populate embedded file content: %v", err)
```
should we go the other way round and return a custom error if ReadFile(path) fails with NotExist?

Nirusu replied:
I have a slight preference for wrapping PopulateContent's error since it's imo not super well positioned in this function and provides some context that we fail inside the file embedding.

Though the error does usually contain the offending filename anyway so it's not a huge pro argument imo.

What would be the advantage of going the other way around? More specific error handling rather than the more generic approach we currently have?

thomasten replied:
Yes, someone may introduce a similar bug in the future by adding something that could return NotExist. But I also get your point. I leave it up to you.

Nirusu replied:
I'll do both I guess. My current change for clarity that something is wrong with the files to be embedded, the other one for more specific error handling.

Nirusu commented:
Okay, switched to wrapped errors and added a specific error type for an expected missing enclave.json.

You can check if it is okay this way. More advanced error handling in Go is a bit fugly to be honest.

thomasten commented on ego/cli/sign.go:
```diff
@@ -134,6 +137,9 @@ func (c *Cli) signExecutable(path string) error {
 // JSON could not be unmarshalled
 func (c *Cli) readConfigJSONtoStruct(path string) (*config.Config, error) {
 	data, err := c.fs.ReadFile(path)
+	if os.IsNotExist(err) {
```
```suggestion
	if errors.Is(err, os.ErrNotExist) {
```
nit: modernize

thomasten commented on ego/cli/sign.go:
```diff
@@ -134,6 +137,9 @@ func (c *Cli) signExecutable(path string) error {
 func (c *Cli) readConfigJSONtoStruct(path string) (*config.Config, error) {
 	data, err := c.fs.ReadFile(path)
+	if os.IsNotExist(err) {
+		return nil, ErrConfigDoesNotExist
```
```suggestion
		return nil, fmt.Errorf("%w: %v", errConfigDoesNotExist, path)
```
A bit unorthodox, but maybe it's helpful for the user to get the path in the error message, especially in the enclave.json case. What do you think?

(I think you would usually define a custom error type for this, but I guess this is what you mean by "more advanced error handling in Go is a bit fugly", hence this proposal.)

thomasten commented on ego/cli/sign_test.go:
```diff
@@ -231,8 +230,25 @@ NumHeapPages=256
+	// Create source file, test should pass
+	content := []byte{2, 0, 3}
+	require.NoError(fs.WriteFile(sourcePath, content, 0))
 	require.NoError(fs.WriteFile("enclave.json", []byte(configJSON), 0))
```
```suggestion
```
by removing this line, we'd have an implicit test for the actual bug

Nirusu replied:
But you would still keep the other tests, right?

thomasten replied:
yes
